### PR TITLE
Add TypeNode#has_method?

### LINKED ISSUE #4472: Add has_method? to TypeNode
As suggested in #4471 it would be nice to have a `has_method?` instance method for `TypeNode` to implement a behaviour similar to `responds_to?` inside a macro.
There are already `has_constant?` and `has_attribute?`, it would make sense to have such a method for instance methods.

It won't be an additional feature for macros, just provide a short cut for ```type_node.methods.any? { |method| method.name == method_name }```.

### PR DESCRIPTION
**Pull request by Sija**
Fixes #4472

### CODE REVIEW

**Comment by bew** [comment]
Nice addition, you should add some specs for this

**Comment by Sija** [comment]
@bew there aren't any for the rest of related methods, I can add it though.

**Comment by bcardiff** [comment]
@Sija yes, please let's add some spec in `/spec/compiler/codegen/macro_spec.cr`.

I would suggest also to split the fixes in the docs in other PR 🙏 

Regarding the functionality, I have some doubts how it should interact with `method_missing`. Currently `responds_to` returns false if no method with that name is invoked somewhere, which is reasonable given that the types are fixed after the compilation finishes. But during macros evaluation, depending the evaluation order the has_method might return true if later a method_missing generates that method.

File: src/compiler/crystal/macros/methods.cr [hunk 1]
```diff
@@ -1533,6 +1538,15 @@ module Crystal
       ArrayLiteral.new(defs)
     end
 
+    def self.has_method?(type, name)
+      type.defs.try &.each do |name, metadatas|
```
Comment (asterite) [inline src/compiler/crystal/macros/methods.cr]: You can simply use the `name` variable here, there's no need to traverse all metadatas

Comment (asterite) [inline src/compiler/crystal/macros/methods.cr]: @Sija Ping :-)

Comment (Sija) [inline src/compiler/crystal/macros/methods.cr]: `BoolLiteral.new !!type.defs[name]?` will do?

Comment (asterite) [inline src/compiler/crystal/macros/methods.cr]: Yes, or `type.defs.has_key?(name)`

Comment (asterite) [inline src/compiler/crystal/macros/methods.cr]: Actually... `type.has_def?(name)`

Comment (Sija) [inline src/compiler/crystal/macros/methods.cr]: done :)

File: src/compiler/crystal/macros/methods.cr [hunk 2]
```diff
@@ -1533,6 +1538,10 @@ module Crystal
       ArrayLiteral.new(defs)
     end
 
+    def self.has_method?(type, name)
+      BoolLiteral.new(!!type.has_def?(name))
```
Comment (bew) [inline src/compiler/crystal/macros/methods.cr]: I think that `BoolLiteral.new(type.has_def?(name))` is enough, as `Type#has_def?` is already returning a boolean

Comment (Sija) [inline src/compiler/crystal/macros/methods.cr]: @bew nope, returned type is `Bool | Nil`

Comment (bew) [inline src/compiler/crystal/macros/methods.cr]: right, sorry about that (found why: `some_hash.try &.has_key?(some_key)` can be `Nil` with the `try`..)

**Comment by Sija** [comment]
@bcardiff I've added specs as per request. I'm not sure what to do regarding `method_missing` and macros evaluation order... any ideas? /cc @asterite

**Comment by asterite** [comment]
@Sija don't worry about that, it's a separate issue

**Comment by Sija** [comment]
Is it GTG? Travis CI xcode8.2 build failed, rest 's good.

**Pull request merged**